[20기 최준식] [Mission 2] 회원가입 엔드포인트

JunsikChoi opened the pull request:
## 수정 사항 간략한 한줄 요약
- User 데이터 모델 생성 및 SignUp View 추가. 
- API 입력 데이터 검증 및 DB 입력시 예외 처리
- 패스워드 / 이메일 검증 로직 추가
- 패스워드 해쉬 후 저장

## 수정 사항들 자세한 내용

### User 데이터 모델 생성 (98db54a)

#### 필수 필드
- email
- password

#### 비 필수 필드
- username
- phone_number

#### 비 중복 필드
- email
- password
- phone_number

### SignUp View 작성
- http 메서드 POST로 제한
- 필수 필드 포함 검증 및 통과 못할 시 KEY_ERROR 반환
- 이메일 및 패스워드 검증 (796ab57) 및 ValidationError 처리 
- 레코드 입력시 중복 예외 처리 (b3db05b, 04e1c91)
- 입력 성공 시 SUCCESS & 생성된 데이터 반환 (패스워드 제외)

### 중복된 데이터 입력 방지 (b3db05b, 04e1c91)
- DuplicatedEntryError 커스텀 에러 클래스 생성
- utils.py의 check_duplicate 통해 모델의 unique 필드 체크 및 해당 필드 중복시 중복 에러 생성
- SignUpView의 post 함수에서 해당 에러 발생시 중복된 필드명 및 409 에러 반환

### User App URLConf 루트 URL Conf에 추가 (6b293fd)

### ADD: User Data Model 마이그레이션 파일 추가 (6545217)

### Users App Setting 파일에 추가 (2386f9e)

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
조건을 만족한다면 각 빈칸에 x 표시 해주세요
- [x ] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [ x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [ x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.

JunsikChoi committed:
MODIFY: Register Users App to Setting

JunsikChoi committed:
ADD: User App URLConf and Register to Root URLConf

JunsikChoi committed:
ADD: User Data Model

JunsikChoi committed:
ADD: User App SignUp View & Email/Password Validator

JunsikChoi committed:
ADD: User Data Model Migration File

Kang-kyunghun commented on students/junsikchoi/users/urls.py:
```diff
@@ -0,0 +1,13 @@
+from django.urls import path
+from users.views import SignInView, SignUpView
+
+urlpatterns = [
+    path(
+        "/signUp",
+        SignUpView.as_view(),
+    ),
+    path(
+        "/signIn",
+        SignInView.as_view(),
+    ),
+]
```
혹시 자동 줄맞춤 기능 사용하셨나요? 오히려 가독성이 떨어져 보입니다.

JunsikChoi replied:
저는 VSCode에서 파이썬 Black Formatter 써서 정렬하고 있습니다. 
위코드 컨벤션 (`일반적으로 한 줄에는 79자를 안넘어가도록 코드를 작성합니다. 한 줄에 79자를 넘어가면 다음줄로 나누어서 작성하는것이 일반적인 가이드 입니다. 하지만 저희 wecode는 79자도 너무 많다고 생각합니다. `)에 맞추려고 1줄 최대 글자수 제한을 기존 80자에서 59자로 제한했더니 너무 빡빡하게 포매팅 된 것 같으니 제한을 넓혀서 다시 포매팅 하겠습니다.

Kang-kyunghun commented on students/junsikchoi/users/validators.py:
```diff
@@ -0,0 +1,21 @@
+
+
+def validate_password(password):
+    min_password_length = 8
```
파이썬에서 상수는 대무자로만 이루어져 있습니다.

JunsikChoi replied:
수정 반영하겠습니다!

Kang-kyunghun commented on students/junsikchoi/users/views.py:
```diff
@@ -1,2 +1,87 @@
+import json
+from django.views import View
+from django.http import JsonResponse
+from users.models import User
+from django.db import IntegrityError
+from django.core.exceptions import ValidationError
+from users.validators import (
+    validate_email,
+    validate_password,
```
위코드 코드 컨벤션 규칙에 맞게 수정해 주세요.

JunsikChoi replied:
수정 반영 하겠습니다!

Kang-kyunghun commented on students/junsikchoi/users/views.py:
```diff
@@ -1,2 +1,87 @@
+from users.validators import (
+    validate_email,
+    validate_password,
+)
```
공백은 1칸만 해주세요.

JunsikChoi replied:
black 포매터에 이 부분 설정하는 부분이 없어 불가피하게 2줄로 맞춰 자동포맷하고 있습니다. 개인적으로 코드 포매팅에 정해진 정답은 없다고 생각하는데, 굳이 코드에 한 줄 아끼기 위해 자동 포매팅 기능을 포기하기는 너무 비효율적이라고 생각해서 이 부분은 black formatter가 자동으로 설정한대로 쓰고자 하는데 크게 문제가 될까요? 제가 없는 스타일을 쓰는게 아니고 자동 포매터 자체도 많은 사람이 써서 그대로 따라가는것이 그렇게 커다란 문제가 될 거라고는 생각치 않아서요.

Kang-kyunghun replied:
당연히 혼자 코딩하는 거면 어떤 포멧으로 가든 상관 없습니다. 그러나 해당 그룹에서 서로 이렇게 사용하고자 하는 약속이 있는데, 그 약속을 혼자 지키지 않는 것은 문제가 된다고 생각합니다.

Kang-kyunghun commented on students/junsikchoi/westagram/urls.py:
```diff
@@ -1,18 +1,3 @@
-from django.urls import path
-
-urlpatterns = []
+urlpatterns = [path("users", include("users.urls"))]
```
오히려 여기는 너무 빡빡하게 해서 가독성이 떨어지네요.
```suggestion
urlpatterns = [
    path("users", include("users.urls"))
]
```

JunsikChoi replied:
음.. 주관적으로 말씀하셔서 조금 헷갈립니다. 기준 글자수를 알려주시면 정확히 설정해서 다시 포매팅 후 올리겠습니다.

Kang-kyunghun commented on students/junsikchoi/users/views.py:
```diff
@@ -1,2 +1,87 @@
-# Create your views here.
+
+class SignUpView(View):
+    # Limit HTTP methods to POST
+    http_method_names = ["post"]
+
+    # Store Model Fields (Required, non-required)
+    required_fields = [
+        _.attname
+        for _ in User._meta.get_fields()
+        if _.blank == False
+    ]
+
+    non_required_fields = [
+        _.attname
+        for _ in User._meta.get_fields()
+        if _.blank == True
+    ]
```
이것들이 필요한 이유와 그래서 왜 사용 했는지 설명해 주세요.

JunsikChoi replied:
이 부분 만들때 view 부분에 처음 request.body를 받아와서 처리할 때 각 모델 클래스가 받아오는 필드 명을 하나씩 하드코딩해서 작성하는 것이 불필요하다고 생각되었습니다. 물론 지금의 User 모델 같은 경우는 딱히 그럴 필요가 없어 보이지만 나중에 많은 필드를 갖는 모델을 처리할 때 쓰기 위해 모델에서 자동으로 필드명과 필드 attribute를 검사해서 분류하는 방법을 찾아 쓴 것입니다.

구글링 해 본 결과 Django 모델에서 _meta attribute를 통해 모델이 가진 필드에 접근이 가능하고 필드에 설정된 blank나 unique 같은 속성값과 필드 이름에도 접근이 가능했습니다. 

음.. 완전히 각 필드명을 하드코딩해준 경우와 비교했을 때 장단점이 있다고 생각하는데, 

장점으로는,
- 모델의 필드명을 바꾸거나 추가해도 view 부분에 이 부분이 달라지지 않아도 됩니다.
- 많은 필드를 갖는 모델에서 필드를 읽어들어와 처리하기 쉽습니다.
- 불필요한 오타의 위험이 없습니다.

단점으로는,
- User 모델과 같이 작은 모델에는 불필요하다고 느껴집니다.
- 코드 가독성이 떨어집니다. (하지만 이 부분은 변수 명 자체가 뒤 로직을 충분히 설명하고 있기에 커버 가능하다고 봅니다.)

가 있다고 생각합니다. 

저는 지금 여기 모델 뿐 아니라 나중에 필드가 많고 하드코딩이 아닌 자동으로 처리가 필요할 때 사용하고 싶어서 위와 같이 코드 작성했습니다. 

list comprehension 안에 `_` 쓴 이유는 굳이 변수명을 줄 필요가 없다고 생각해서 입니다.

JunsikChoi replied:
+ 추가로 저는 운산님이 작성하신 예외처리와 같이 하면, (https://github.com/wecode-bootcamp-korea/20-Westagram-backend/blob/8c55975b6afeea71a6b501b35ae240d0e8a574b9/students/unsan-Jeong/users/views.py#L11-L38) objects.create의 age = data['age'] 부분에서 data에 'age' 키가 존재하지 않는 경우를 없애기 위해 저렇게 작성했는데, 불필요하게 코드가 복잡해진것같습니다~ 다음번엔 가독성 좋게 짜보도록 하겠습니당.

Kang-kyunghun replied:
일단 내일 더 자세히 이야기 하겠지만, 왜 필드명과 key 값이 항상 같을 거라고 생각하시는 거죠? 당연히 최대한 비슷하게 가려고 하겠지만, key값은 프론트에서 주는 값이고 이 값은 항상 필드명과 같을 수는 없습니다. 그리고 꼭 필드에 들어가야 하는 값이 들어오는 것도 아니고요.

커버가 가능하지 안을 정도로 가독성이 너무 떨어집니다. 같은 기능을 하는 운산님 코드와 준식님 코드가 현업을 진행하는데 있어 더 좋은 코드일까요? 

기슬적으로는 인정하는 부분이지만, 단지 오버 엔지니어링으로 밖에 보이지 않습니다.

Kang-kyunghun commented:
일단 코드가 전체적으로 가독성이 떨어집니다. 내일 오전에 저랑 잠깐 만나겠습니다. 그 전까지 리뷰 사항 반영해 주세요.

JunsikChoi committed:
FIX: CODE FORMATS

- Increase Code Readability
- Change import order in views.py
- Change number of empty line

JunsikChoi commented:
변경사항 반영 완료!

JunsikChoi committed:
MODIFY: Add More Specific Error Class

JunsikChoi committed:
MODIFY: SignUpView Class

- Increase code Readability
- Modify view logic for post request
- Exception handling for Invalid Email/Password, Undecodable JSON, Empty
required field, duplicated field, and other DB Integrity Error

JunsikChoi committed:
ADD: Add utils.py

- Add custom Exception for duplicated entry
- duplicate check function that raises DuplicatedEntryError

JunsikChoi commented:
SignUp View 처리 로직 수정 완료!

JunsikChoi committed:
MODIFY: null=True for Non-Required Fields

Changed field option of username and phone_number to null=True

blank=True stores empty non-required field record to empty string by
default.

This makes Integrity error when another empty non-required field entry
comes in.

JunsikChoi committed:
MODIFY: User Field Options

JunsikChoi committed:
MODIFY: Change SignUpView Data Access Method

Prior method by dictionary key accessing occurs KeyError when Key is not
exists.

This makes User.objects.create(<non-required-field>:
data[<non-required-field>] ...) impossible when non required field
actually not provided.

Instead, by using data.get(<non-required-field>), I can bypass the
problem, because dictionary.get(key) does not make KeyError but returns
None when key is not exists.

JunsikChoi commented:
비필수 필드의 경우 blank=True로 설정했을 시, 빈 필드값으로 레코드를 생성하면 디폴트로 ''(빈스트링)이 들어갑니다.
phone_number와 username은 비 필수 필드이자 unique 필드인데 빈 스트링도 중복에러를 발생시키기에 IntegrityError Exception를 발생시키는 문제가 있습니다. 그래서 null=True로 설정해서 빈 필드값이 들어가면 NULL 값이 입력되게 했습니다.

또,  data 접근 방식을 키 방식으로 하면 비 필수 필드가 주어지지 않았을 때 딕셔너리 키에러를 발생시켜서, 키 접근방식 아닌 dict.get() 접근 방식으로 바꿨습니다.

Kang-kyunghun commented on students/junsikchoi/users/validators.py:
```diff
@@ -0,0 +1,21 @@
+def validate_email(email):
+    required_character = ["@", "."]
+
+    for chracter in required_character:
+        if chracter not in email:
+            raise ValidationError(
+                f"{email} is not an valid email.",
+                params={"value": email},
+            )
```
결국은 if를 for문으로 돌려서 확인하는데, 2개의 인자에 대해 확인하는데 for문까지 굳이 가야 하나라는 생각이 듭니다. 만약 확장성을 고려 하신거라면 정규표현식으로 가는 것이 더 좋을 거 같습니다.

JunsikChoi replied:
알겠습니다.

Kang-kyunghun commented on students/junsikchoi/users/views.py:
```diff
@@ -1,2 +1,66 @@
+import json
+from json import JSONDecodeError
+from django.db import IntegrityError
+from django.views import View
+from django.http import JsonResponse
+from django.core.exceptions import ValidationError
+from users.models import User
+from users.validators import (
+    validate_email,
+    validate_password,
+)
```
위코드 코드 컨벤션을 지겨주세요.
```suggestion
import json
from json                   import JSONDecodeError

from django.db              import IntegrityError
from django.views           import View
from django.http            import JsonResponse
from django.core.exceptions import ValidationError

from users.models           import User
from users.validators       import validate_email, validate_password
```

JunsikChoi replied:
넵!

Kang-kyunghun commented on students/junsikchoi/users/validators.py:
```diff
@@ -0,0 +1,21 @@
+
+    if len(str(password)) < MIN_PASSWORD_LENGTH:
+        raise ValidationError(
+            "Your Password is too short. Use Longer Password."
```
- 리턴 메시지는 클라이언트가 보는 것이 아닌고 프론트엔드가 보는 메시지 입니다. 굳이 문장으로 안 하시고, 간단하게 표현해도 좋을 것 같습니다.
- 메시지 형식은 통일하는 것이 좋습니다. 다 대문자로 썼으면 대문자로만, _을 사용해서 단어을 이였다면 계속 _을 사용하는 것처럼요

JunsikChoi replied:
확인했습니다. 통일하겠습니다~

Kang-kyunghun commented on students/junsikchoi/users/views.py:
```diff
@@ -1,2 +1,66 @@
+        except DuplicatedEntryError as e:
+            return JsonResponse({"message": e.err_message}, status=409)
+
+        except IntegrityError as e:
+            return JsonResponse(
+                {
+                    "message": "DataBase Integrity Error: {}".format(str(e.__cause__)),
+                },
+                status=500,
+            )
```
500 에러를 보내는 것은 제대로 예외처리가 되지 않았다는 의미를 가지고 있는데, 이것을 일부로 주게 되면, 디버깅이 힘들어지고, 스스로 예외처리 못 했다라고 말하는 상황이 될 것 같습니다.

JunsikChoi replied:
음 그럼 윗부분에서 중복처리를 했으므로 IntegrityError except 문은 없애도 될까요?
IntegrityError가 정확이 어떤 데이터베이스 예외 상황에서 발생하는지 찾아봤는데 정확한게 없더라구요.
그래서 중복이 해결 된 상황에서 나머지 상황은 500 에러로 처리했었습니다.

Kang-kyunghun replied:
IntegrityError 에러가 DB에서 나는 에러에 대해서 나타내는 예외처리 입니다. 중복에 대한 것도 unique처리가 되어 있고, 이는  DB에 데이터를 저장할 때 문제가 발생해서 IntegrityError를 띄웁니다.  로직상 중복에 대한 예외처리를 했기 때문에 지워도 무방하다고 봅니다.

Kang-kyunghun commented on students/junsikchoi/users/views.py:
```diff
@@ -1,2 +1,66 @@
 
-# Create your views here.
+class SignUpView(View):
+    # Limit HTTP methods to POST
+    http_method_names = ["post"]
```
이 부분은 저희가 View class를 상속 받고, urls에서 as_view()을 사용함에 따라 이미 처리된 부분 입니다.

JunsikChoi replied:
앗 그럼 as_view() 로 처리하고 클래스 내에 get 메서드가 없을 경우 자동으로 예외 처리가 되나보네요 확인했습니다.

Kang-kyunghun replied:
네 맞습니다.없는 method로 통신하게 되면 405 에러 발생합니다.

JunsikChoi committed:
MODIFY: Delete http_method_names

JunsikChoi committed:
MODIFY: Delete comments

JunsikChoi committed:
MODIFY: Delete IntegrityError Exception

JunsikChoi committed:
MODIFY: Simplify email validator logic

JunsikChoi committed:
MODIFY: Import Statement Code Formatting

JunsikChoi committed:
MODIFY: Unify JsonResponse Format

JunsikChoi commented:
반영 완료했습니다!

Kang-kyunghun commented on students/junsikchoi/users/views.py:
```diff
@@ -1,2 +1,44 @@
+            
+            check_duplicate(User, data)
+
+            user = User.objects.create(
+                email = data.get('email'),
+                password = data.get('password'),
+                username = data.get('username'),
+                phone_number = data.get('phone_number'),
+            )
```
https://www.notion.so/wecode/wecode-Python-Coding-Convention-0bd47a6018634454b9ee1e27984cb83a
위코드 코드 컨벤션 내용입니다. 참고해서 수정해 주세요.

Kang-kyunghun commented on students/junsikchoi/utils.py:
```diff
@@ -0,0 +1,19 @@
+    non_duplicatable_fields = [
+        _.attname
+        for _ in model._meta.get_fields()
+        if _.unique == True
+    ]
```
- .unique의 결과가 boolean이라면 == True을 제외 시키는 것이 조금 더 파이썬 다운 코드라고 생각됩니다.

JunsikChoi committed:
MODIFY: Add Password Hashing in SignUpView

JunsikChoi committed:
MODIFY: DB Migration. Longer Field for Hashed Password

JunsikChoi committed:
FIX: Code Formatting

Kang-kyunghun commented:
LGTM~~!
